server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java

```diff
@@ -141,53 +143,15 @@ private TimeSeriesIdFieldMapper() {
     public void postParse(DocumentParserContext context) throws IOException {
         assert fieldType().isIndexed() == false;
 
-        // SortedMap is expected to be sorted by key (field name)
-        SortedMap<BytesRef, BytesReference> dimensionFields = context.doc().getDimensionBytes();
-        BytesReference timeSeriesId = buildTsidField(dimensionFields);
-        context.doc().add(new SortedDocValuesField(fieldType().name(), timeSeriesId.toBytesRef()));
-    }
-
-    public static BytesReference buildTsidField(SortedMap<BytesRef, BytesReference> dimensionFields) throws IOException {
-        if (dimensionFields == null || dimensionFields.isEmpty()) {
-            throw new IllegalArgumentException("Dimension fields are missing.");
-        }
-
-        try (BytesStreamOutput out = new BytesStreamOutput()) {
-            encodeTsid(out, dimensionFields);
-            BytesReference timeSeriesId = out.bytes();
-            if (timeSeriesId.length() > LIMIT) {
-                throw new IllegalArgumentException(NAME + " longer than [" + LIMIT + "] bytes [" + timeSeriesId.length() + "].");
-            }
-            return timeSeriesId;
-        }
+        TimeSeriesIdBuilder timeSeriesIdBuilder = (TimeSeriesIdBuilder) context.doc().getDimensions();
+        context.doc().add(new SortedDocValuesField(fieldType().name(), timeSeriesIdBuilder.build().toBytesRef()));
```

Comment:
I think I would either 1) move this logic to LuceneDocument and teach it how to create tsid or 2) move the builder to context and keep LuceneDocument blissfully unaware of time series world otherwise this "Please hold this for me" type of interaction with LuceneDocument is a bit weird. I think I would prefer 2). The first solution would make more sense if we wanted to do keep different dimensions for nested documents, but at the moment they share the same set of dimensions anyway (which is probably a topic for another discussion).

Reply:
I'll move it.

Reply:
It's so much cleaner!